torch/csrc/autograd/variable.cpp

```diff
@@ -115,6 +123,90 @@ void Variable::set_data(const at::Tensor &new_data) {
   get()->shallow_copy_from(new_data.getIntrusivePtr());
 }
 
+static void set_refcount(at::TensorImpl *impl, size_t cur_refcount, size_t target_refcount, bool is_weak_count) {
+  if (target_refcount > cur_refcount) {
+    for (size_t i = cur_refcount; i < target_refcount; i++) {
+      if (!is_weak_count) {
+        c10::raw::intrusive_ptr::incref(impl);
+      } else {
+        c10::raw::weak_intrusive_ptr::incref(impl);
+      }
+    }
+  } else {
+    for (size_t i = cur_refcount; i > target_refcount; i--) {
+      if (!is_weak_count) {
+        c10::raw::intrusive_ptr::decref(impl);
+      } else {
+        c10::raw::weak_intrusive_ptr::decref(impl);
+      }
+    }
+  }
+}
+
+void Variable::_set_data_swap_impl(const at::Tensor &new_data) {
+  AT_ASSERT(new_data.getIntrusivePtr().use_count() == 1);
+
+  // Resets gradient accumulator if metadata is out of date
+  reset_grad_accumulator(new_data.device(), new_data.type());
+
+  // We change this `Variable`'s TensorImpl, but preserves its `pyobj_` pointerc,
+  // so that previous references to this `Variable` in Python are still valid.
+  auto new_impl = new_data.unsafeGetTensorImpl()->shallow_copy_and_detach(
+    /*version_counter=*/get()->version_counter(),
+    /*allow_tensor_metadata_change=*/get()->allow_tensor_metadata_change());
+  new_impl->set_autograd_meta(std::move(get()->detach_autograd_meta()));
+  new_impl->set_pyobj(get()->pyobj());
+  size_t impl_refcount_saved = impl_.use_count();
+  size_t impl_weakcount_saved = impl_.weak_use_count();
+  size_t new_impl_refcount_saved = new_impl.use_count();
+  size_t new_impl_weakcount_saved = new_impl.weak_use_count();
+  std::swap(*(impl_.get()), *(new_impl.get()));
+  set_refcount(impl_.get(), impl_.use_count(), impl_refcount_saved, false);
+  set_refcount(impl_.get(), impl_.weak_use_count(), impl_weakcount_saved, true);
+  set_refcount(new_impl.get(), new_impl.use_count(), new_impl_refcount_saved, false);
+  set_refcount(new_impl.get(), new_impl.weak_use_count(), new_impl_weakcount_saved, true);
```

Comment:
@gchanan @ezyang I would like to suggest an alternative way to approach this problem: previously we discussed that changing `impl_` of a Tensor would break all previous references to the same Tensor in C++. However, if we only swap the TensorImpl content of the instrusive_ptr `impl_`, the intrusive_ptr can stay unchanged, and all previous references to the same Tensor will still work. (I added tests in https://github.com/pytorch/pytorch/pull/21048/files#diff-7a22eab7d8855560eec34e0e8f0aad18R311 to validate that this is the case.)

Reply:
This is one way to guarantee backward compatibility of `model.to(device)`, and it also allows us to implement `tensor.to_()`, which can be used in `model.to()` to correctly track the version counter.

Reply:
This PR still needs quite a bit of clean up, especially regarding where `to_()` is defined in C++, but I'd love to have a discussion about the general approach first before cleaning it up.

Reply:
Changing the `impl_` of `Tensor` is observationally indistinguishable from assigning over the Tensor, e.g., `x = y`. You'll see everything that is a reference to the same Tensor get updated, but if anyone has a new Tensor with a refcount bump (e.g., via `Tensor new_tensor = old_tensor`), it won't be updated.

Reply:
@ezyang Is it the same case as https://github.com/pytorch/pytorch/pull/21048/files#diff-7a22eab7d8855560eec34e0e8f0aad18R314, where we do `torch::Tensor tensor1 = torch::randn({3, 4}).to(torch::kFloat);` and then `torch::Tensor tensor2 = tensor1;`? The test seems to work even if `tensor2` causes a refcount bump.

Reply:
I read your diff and it looks like I misinterpreted your explanation of what you were doing. I see your "swap impl_" implemented as:

```
  std::swap(*(impl_.get()), *(new_impl.get()));
```

this means swap is taking `TensorImpl&` as its argument here; aka you are using move assignment/construction to swap the literal contents of TensorImpl. I agree that this is consistent with the behavior you see. But it will not work if you have subclasses of TensorImpl.